maven-core/src/main/java/org/apache/maven/DefaultMaven.java

```diff
@@ -463,51 +484,68 @@ private void validateLocalRepository( MavenExecutionRequest request )
         }
     }
 
-    private Collection<AbstractMavenLifecycleParticipant> getLifecycleParticipants( Collection<MavenProject> projects )
+    private <T> Collection<T> getExtensionComponents( Collection<MavenProject> projects, Class<T> role )
     {
-        Collection<AbstractMavenLifecycleParticipant> lifecycleListeners = new LinkedHashSet<>();
+        Collection<T> foundComponents = new LinkedHashSet<>();
 
         ClassLoader originalClassLoader = Thread.currentThread().getContextClassLoader();
         try
         {
             try
             {
-                lifecycleListeners.addAll( container.lookupList( AbstractMavenLifecycleParticipant.class ) );
+                foundComponents.addAll( container.lookupList( role ) );
             }
             catch ( ComponentLookupException e )
             {
                 // this is just silly, lookupList should return an empty list!
-                logger.warn( "Failed to lookup lifecycle participants: " + e.getMessage() );
+                logger.warn( "Failed to lookup " + role + ": " + e.getMessage() );
             }
 
-            Collection<ClassLoader> scannedRealms = new HashSet<>();
+            foundComponents.addAll( getProjectScopedExtensionComponents( projects, role ) );
+        }
+        finally
+        {
+            Thread.currentThread().setContextClassLoader( originalClassLoader );
```

Comment:
Why is this necessary now, the TCCL is never overridden here?

Reply:
I think this is obsolete because its done in the method call already, I'll adjust this.